### PR TITLE
Dc sync eikon

### PR DESCRIPTION
**Pull request by lielShraga**

### CODE REVIEW

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 1]
```diff
@@ -845,4 +880,14 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
         .withAttributes(ActorAttributes.supervisionStrategy(localDecider))
     syncingEngine
   }
+
 }
+
+
+//TODO:
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: remove

File: server/cmwell-dc/src/universal/conf/application.conf [hunk 2]
```diff
@@ -54,11 +54,15 @@ metrics {
...
 irwServiceDao {
-  hostName = "notRelevant" #localhost
-  clusterName = "notRelevant" #cmwell
+  hostName = "127.0.0.1" #localhost
+  clusterName = "cm-well-p" #cmwell
   keySpace2 = data2
 }
 
+fingerprint{
+  threadsParallelism = 5
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/universal/conf/application.conf]: done

File: server/cmwell-dc/src/pack/application.conf [hunk 3]
```diff
@@ -76,5 +76,9 @@ irwServiceDao {
   keySpace2 = data2
 }
 
+fingerprint{
+  threadsParallelism = 5
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/pack/application.conf]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/TsvRetriever.scala [hunk 4]
```diff
@@ -65,7 +65,7 @@ object TsvRetriever extends LazyLogging {
     logger.trace(
       s"parseTSVAndCreateInfotonDataFromIt: [path='$path',uuid='${uuid.utf8String}',idxt='$idxt']"
     )
-    InfotonData(InfotonMeta(path, uuid, idxt), empty)
+    InfotonData(InfotonMeta(path, Some(uuid), Some(idxt)), empty)
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/TsvRetriever.scala]: for now I'm removing the opt and send hardcoded uuid and indextime

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/MessagesTypesAndExceptions.scala [hunk 5]
```diff
@@ -24,15 +24,27 @@ import scala.util.{Failure, Success}
   */
 object MessagesTypesAndExceptions {
 
-  case class DcInfoKey(id: String, location: String, transformations: Map[String, String]) {
+  case class DcInfoKey(id: String, location: String, transformations: Map[String, String], ingestOperation:String) {
     override def toString: String =
       s"[id: $id, location: $location, transformations: ${transformations.mkString("(", ",", ")")}]"
   }
+
+  abstract sealed class DcInfoExtra
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/MessagesTypesAndExceptions.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/MessagesTypesAndExceptions.scala [hunk 6]
```diff
@@ -24,15 +24,27 @@ import scala.util.{Failure, Success}
   */
 object MessagesTypesAndExceptions {
 
-  case class DcInfoKey(id: String, location: String, transformations: Map[String, String]) {
+  case class DcInfoKey(id: String, location: String, transformations: Map[String, String], ingestOperation:String) {
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/MessagesTypesAndExceptions.scala]: will do in phase2

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/SingleMachineInfotonIngester.scala [hunk 7]
```diff
@@ -48,7 +48,7 @@ object SingleMachineInfotonIngester extends LazyLogging {
   private val dcaToken = Settings.dcaUserToken
   type IngestState = (IngestInput, IngestStateStatus)
 
-  def apply(dcKey: DcInfoKey, location: String, decider: Decider)(
+  def apply(dcKey: DcInfoKey, location: String, decider: Decider, ingestOperation:String)(
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/SingleMachineInfotonIngester.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonAllMachinesDistributerAndIngester.scala [hunk 8]
```diff
@@ -65,7 +66,7 @@ object InfotonAllMachinesDistributerAndIngester extends LazyLogging {
           val (host, portOpt) = hosts(i)
           val location = s"$host:${portOpt.getOrElse(80)}"
           val ingestFlow =
-            SingleMachineInfotonIngester(dckey, location, decider)
+            SingleMachineInfotonIngester(dckey, location, decider, ingestOperation)
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonAllMachinesDistributerAndIngester.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 9]
```diff
@@ -15,35 +15,29 @@
...
+import com.typesafe.config.ConfigFactory
+import k.grid.GridReceives
+import play.api.libs.json.{JsValue, _}
 
+import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent._
-import ExecutionContext.Implicits.global
 import scala.concurrent.duration._
-import scala.util.matching.Regex
+import scala.language.reflectiveCalls
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 10]
```diff
@@ -143,6 +138,10 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
   //context.system
   implicit val sys = ActorSystem("stream-dc")
   implicit val mat = ActorMaterializer(ActorMaterializerSettings(sys))
+  implicit val scheduler = sys.scheduler
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: not related to fp code

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 11]
```diff
@@ -143,6 +138,10 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
...
+  lazy val dao = Dao(Settings.irwServiceDaoClusterName, Settings.irwServiceDaoKeySpace2, Settings.irwServiceDaoHostName, 9042, initCommands = None)
+  lazy val zStore : ZStore = ZStore.apply(dao)
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 12]
```diff
@@ -218,6 +217,10 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
         //The sync was completed successfully. Save the position key for the next sync to start.
         val newSyncMap: SyncMap = currentSyncs - dcInfo.key + (dcInfo.key -> SyncerDone(dcInfo.positionKey.get))
         currentSyncs = newSyncMap
+        if(dcInfo.dcInfoExtraType.isDefined && dcInfo.dcInfoExtraType.get == "fingerprint")
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done.

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 13]
```diff
@@ -308,15 +311,27 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
     }
   }
 
+  private def retrievePosition(dcType:Option[String], position:String) = {
+    logger.info("lala, web service thread=" + Settings.fingerprintParallelism)
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 14]
```diff
@@ -308,15 +311,27 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
...
+    dcType match{
+      case Some(dType) if dType == "fingerprint" =>
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 15]
```diff
@@ -477,8 +503,12 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
               yield str).mkString("&")
             val qpAndWhStrFinal =
               if (qpAndWhStr.length == 0) "" else "?" + qpAndWhStr
-            val dcKey = DcInfoKey(s"$dataCenterId$qpAndWhStrFinal", location, transformations)
-            DcInfo(dcKey, fromIndexTime, tsvFile = tsvFile)
+            val (dcInfoExtra, ingestOp) = dcInfoExtraType match {
+              case Some(t) if t == "fingerprint" => (FingerPrintJsonParser.extractFingerPrintInfo(f), "_in")
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 16]
```diff
@@ -477,8 +503,12 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
...
+              case _ => (None, "_ow")
+          }
+            val dcKey = DcInfoKey(s"$dataCenterId$qpAndWhStrFinal", location, transformations, ingestOp)
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: phase 2

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+
+import scala.concurrent.ExecutionContext
+
+object FingerPrintFlow {
+
+   def fingerprintFlow(dcInfo: DcInfo)
+                      (implicit ec:ExecutionContext, scheduler:Scheduler, mat:ActorMaterializer, system:ActorSystem) = {
+    Flow[InfotonData]
+      .map(infoton => infoton.data)
+      .via(Framing.delimiter(delimiter = ByteString("\n", "UTF-8"), 20000,
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+        allowTruncation = false))
+      .map(_.utf8String)
+      .filter(_.contains("<http://graph.link/ees/Uuid>"))
+      .map(triple => triple.split(" ")(2).dropRight(1).drop(1))
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala]: correct. remove the " "

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+      .scan(Set.empty[String])((total, element) => total.+(element))
+      .mapConcat(identity)
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+       .mapAsync(Settings.fingerprintParallelism)(uuid => FingerPrintWebService.generateFingerPrint(fpUrl(dcInfo, uuid), getDestinationCluster(dcInfo)))
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+      .filter(_.nonEmpty)
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  def fpUrl(dcInfo: DcInfo, uuid: String):String = {
+    dcInfo.dcInfoExtra match {
+      case Some(FingerPrintData(wsCluster, _)) => s"http://$wsCluster/user/$uuid"
+      case _ => throw new IllegalArgumentException("Web service url for fingerprint was not provided")
+    }
+  }
+
+  def getDestinationCluster(dcInfo: DcInfo):String = {
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+      .filter(_.contains("<http://graph.link/ees/Uuid>"))
+      .map(triple => triple.split(" ")(2).dropRight(1).drop(1))
+      .scan(Set.empty[String])((total, element) => total.+(element))
+      .mapConcat(identity)
+       .mapAsync(Settings.fingerprintParallelism)(uuid => FingerPrintWebService.generateFingerPrint(fpUrl(dcInfo, uuid), getDestinationCluster(dcInfo)))
+      .filter(_.nonEmpty)
+      .map(rdf => InfotonData(InfotonMeta(rdf.utf8String.split(" ")(0)), rdf))
+  }
+
+  def fpUrl(dcInfo: DcInfo, uuid: String):String = {
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintFlow.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonAllMachinesDistributerAndIngester.scala [hunk 24]
```diff
@@ -51,7 +51,8 @@ object InfotonAllMachinesDistributerAndIngester extends LazyLogging {
 
   def apply(dckey: DcInfoKey,
             hosts: Vector[(String, Option[Int])],
-            decider: Decider)(implicit sys: ActorSystem, mat: Materializer) = {
+            decider: Decider,
+            ingestOperation:String)(implicit sys: ActorSystem, mat: Materializer) = {
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/InfotonAllMachinesDistributerAndIngester.scala]: done

File: server/cmwell-dc/src/universal/conf/application.conf [hunk 25]
```diff
@@ -54,11 +54,15 @@ metrics {
 }
 
 irwServiceDao {
-  hostName = "notRelevant" #localhost
-  clusterName = "notRelevant" #cmwell
+  hostName = "127.0.0.1" #localhost
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/universal/conf/application.conf]: do u mean that I need to remove it and remain the "NotReleavant"?

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala (hunk 1 above)
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/FingerPrintEngine.scala [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+
+import akka.actor.{ActorSystem, Scheduler}
+import akka.stream._
+import akka.stream.scaladsl.{RunnableGraph, Sink}
+import cmwell.dc.stream.MessagesTypesAndExceptions._
+import cmwell.dc.stream.akkautils.ConcurrentFlow
+import cmwell.dc.stream.fingerprint.FingerPrintFlow
+import cmwell.dc.{LazyLogging, Settings}
+
+import scala.concurrent.ExecutionContext.Implicits.global
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/FingerPrintEngine.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/FingerPrintEngine.scala [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+import scala.language.reflectiveCalls
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/FingerPrintEngine.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/FingerPrintEngine.scala [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+class FingerPrintEngine(dstServersVec: Vector[(String, Option[Int])],
+                            manualSyncList: Option[String])
+    extends LazyLogging {
+
+  import DataCenterSyncManager._
+
+  //todo: should I use the same ActorSystem or create a new one?
+
+  //map: data center ID -> syncer status (warm up, done or running with the materialized value)
+  type SyncMap = Map[DcInfoKey, SyncerStatus]
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/FingerPrintEngine.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/FingerPrintEngine.scala [hunk 29]
```diff
@@ -0,0 +1,100 @@
...
+            )
+        }
+        .withAttributes(ActorAttributes.supervisionStrategy(localDecider))
+    syncingEngine
+  }
+
+}
+
+
+//TODO:
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/FingerPrintEngine.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/FingerPrintEngine.scala [hunk 30]
```diff
@@ -0,0 +1,100 @@
...
+import cmwell.dc.stream.akkautils.ConcurrentFlow
+import cmwell.dc.stream.fingerprint.FingerPrintFlow
+import cmwell.dc.{LazyLogging, Settings}
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.language.reflectiveCalls
+
+
+class FingerPrintEngine(dstServersVec: Vector[(String, Option[Int])],
+                            manualSyncList: Option[String])
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/FingerPrintEngine.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 31]
```diff
@@ -0,0 +1,111 @@
...
+  def toRDF(data:Source[ByteString, _], cluster:String)(implicit ec: ExecutionContext,mat:ActorMaterializer) = {
+    //TODO:remove the filter of "Nan after Yaron's fix
+    data
+      .via(lineSeparatorFrame)
+      .map(fp => fp.utf8String)
+      .map(fp => fp.replace("\\", "\\\\"))
+      .filterNot(x => x.contains("NaN"))
+      .map(fp => FingerPrintData((Json.parse(fp) \ "account_info" \ "UUID").as[String], fp.replace("\"", "\\\"")))
+      .map(fpData => {
+        val subject = s"<http://$cluster/graph.link/ees/FP-${fpData.uuid}>"
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 32]
```diff
@@ -0,0 +1,111 @@
...
+  case class FingerPrintData(uuid:String, data:String)
+  val lineSeparatorFrame = Framing.delimiter(delimiter = endl,
+    200000,
+    allowTruncation = true)
+
+
+
+  def generateFingerPrint(url:String, cluster:String)(implicit ec:ExecutionContext, scheduler:Scheduler,
+                                      mat:ActorMaterializer, system:ActorSystem) = {
+    sendHttpRequest(url)
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 33]
```diff
@@ -0,0 +1,111 @@
...
+    .map(res=> toRDF(res.entity.withSizeLimit(1000000).dataBytes, cluster)).flatten
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 34]
```diff
@@ -0,0 +1,111 @@
...
+import scala.util.{Failure, Success}
+
+case class WebServiceConnectException(message: String) extends  Exception(message)
+object FingerPrintWebService extends LazyLogging{
+
+  val endl = ByteString("\n", "UTF-8")
+  case class FingerPrintData(uuid:String, data:String)
+  val lineSeparatorFrame = Framing.delimiter(delimiter = endl,
+    200000,
+    allowTruncation = true)
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 35]
```diff
@@ -0,0 +1,111 @@
...
+//    cmwell.util.concurrent.retry(5, 10.seconds)(sendRequest(url))(ec)
+  }
+
+
+  private def sendRequest(url: String)(implicit ec:ExecutionContext, mat:ActorMaterializer, system:ActorSystem) = {
+    val webServiceRes = Http(system).singleRequest(HttpRequest(uri = url))
+    val randomId = java.util.UUID.randomUUID.toString
+    val futureResponse = webServiceRes.flatMap(res =>
+      if (res.status.isSuccess()) Future.successful(res)
+      else if(res.status.intValue() == 503) Future.failed(WebServiceConnectException(s"Got 503 error status code, headers=${res.headers}"))
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: done.using Gilad http library

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 36]
```diff
@@ -0,0 +1,111 @@
...
+
+  private def retry[T](delay: FiniteDuration, retries: Int = -1)(task: => Future[HttpResponse])(implicit ec: ExecutionContext, scheduler: Scheduler):
+  Future[HttpResponse] = {
+    task.recoverWith {
+      case e: WebServiceConnectException =>
+        logger.error("Failed to request fingerprint web service,", e.printStackTrace())
+        logger.error("Going to retry till web service will be available")
+        akka.pattern.after(delay, scheduler)(retry(delay)(task))
+      case e: Throwable if retries > 0 =>
+        logger.error("Failed to request web service, need to write to RED log the bulk", e)
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 37]
```diff
@@ -0,0 +1,111 @@
...
+    //TODO:remove the filter of "Nan after Yaron's fix
+    data
+      .via(lineSeparatorFrame)
+      .map(fp => fp.utf8String)
+      .map(fp => fp.replace("\\", "\\\\"))
+      .filterNot(x => x.contains("NaN"))
+      .map(fp => FingerPrintData((Json.parse(fp) \ "account_info" \ "UUID").as[String], fp.replace("\"", "\\\"")))
+      .map(fpData => {
+        val subject = s"<http://$cluster/graph.link/ees/FP-${fpData.uuid}>"
+        val fpInfotonMetaData = subject + " <http://" + cluster + "/meta/sys#type> \"FileInfoton\" .\n" +
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/MessagesTypesAndExceptions.scala [hunk 38]
```diff
@@ -24,15 +24,27 @@ import scala.util.{Failure, Success}
...
     override def toString: String =
       s"[id: $id, location: $location, transformations: ${transformations.mkString("(", ",", ")")}]"
   }
+
+  abstract sealed class DcInfoExtra
+  case class FingerPrintData(webServiceCluster: String, destinationCluster:String) extends DcInfoExtra
+
   case class DcInfo(key: DcInfoKey,
+                    dcInfoExtraType: Option[String],
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/MessagesTypesAndExceptions.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/MessagesTypesAndExceptions.scala [hunk 39]
```diff
@@ -24,15 +24,27 @@ import scala.util.{Failure, Success}
...
+                    dcInfoExtra: Option[DcInfoExtra],
                     idxTime: Option[Long] = None,
                     positionKey: Option[String] = None,
                     tsvFile: Option[String] = None)
 
+
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/MessagesTypesAndExceptions.scala]: what?

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintJsonParser.scala [hunk 40]
```diff
@@ -0,0 +1,34 @@
...
+  */
+package cmwell.dc.stream.fingerprint
+
+import cmwell.dc.stream.MessagesTypesAndExceptions.FingerPrintData
+import play.api.libs.json.{JsArray, JsDefined, JsLookupResult, JsObject}
+
+object FingerPrintJsonParser {
+
+
+  def extractFingerPrintInfo(f: JsLookupResult):Option[FingerPrintData] = {
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintJsonParser.scala]: problem. in a case of none it doesn't compile....

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 41]
```diff
@@ -0,0 +1,111 @@
...
+
+  val endl = ByteString("\n", "UTF-8")
+  case class FingerPrintData(uuid:String, data:String)
+  val lineSeparatorFrame = Framing.delimiter(delimiter = endl,
+    200000,
+    allowTruncation = true)
+
+
+
+  def generateFingerPrint(url:String, cluster:String)(implicit ec:ExecutionContext, scheduler:Scheduler,
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: done

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 42]
```diff
@@ -0,0 +1,111 @@
...
+      else {
+        val blank = ByteString("", "UTF-8")
+        res.entity.dataBytes.runFold(blank)(_ ++ _).map(_.utf8String)
+          .onComplete {
+            case Success(msg) =>
+              logger.error("Message body id=" + randomId + " message=" + msg)
+            case Failure(e) =>
+              logger.error("Got a failure during print entity body, Message body id=" + randomId, e)
+          }
+        Future.failed(new Throwable(s"Got post error code," + res.status + ", headers=" + res.headers +
```
[the remark this answers is not in the document]
Comment (lielShraga) [reply server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: what?

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 43]
```diff
@@ -14,98 +14,73 @@
...
-        val fpInfotonMetaData = subject + " <http://" + cluster + "/meta/sys#type> \"FileInfoton\" .\n" +
-          subject + " <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://graph.link/ees/type/FingerPrint> .\n" +
-          subject + " <http://" + cluster + "/meta/sys#data> \""
-       fpInfotonMetaData + fpData.data + "\" ."
-      })
-      .map(rdfStr => ByteString(rdfStr))
-      .runFold(ByteString.empty){case (acc, fp) => acc ++ fp}
+
+  def toRDF(data:String)(implicit ec:ExecutionContext) = {
+    Future {
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: remove the Future

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 44]
```diff
@@ -14,98 +14,73 @@
...
+      //TODO:remove the filter of "Nan after Yaron's fix
+      val data2 = data.split("\n")(0).replace("\\", "\\\\")
+      val uuid = (Json.parse(data2) \ "account_info" \ "UUID").as[String]
+      val subject = s"<http://graph.link/ees/FP-$uuid>"
+      val rdf2 =
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: 2...

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 45]
```diff
@@ -14,98 +14,73 @@
...
+        s"""$subject <cmwell://meta/sys#type> \"FileInfoton\" .
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: `\` is not needed

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 46]
```diff
@@ -14,98 +14,73 @@
...
+           |$subject  <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://graph.link/ees/type/FingerPrint> .
+           |$subject  <cmwell://meta/sys#data> \"${data2.replace("\"", "\\\"")}\" .
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: \ is not needed

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala [hunk 47]
```diff
@@ -14,98 +14,73 @@
...
+       """.stripMargin
+      ByteString(rdf2)
+    }
 
   }
 
 }
+
+//object FingerPrintWebServiceMain extends App{
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/fingerprint/FingerPrintWebService.scala]: remove

File: server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala [hunk 48]
```diff
@@ -218,6 +216,9 @@ class DataCenterSyncManager(dstServersVec: Vector[(String, Option[Int])],
         //The sync was completed successfully. Save the position key for the next sync to start.
         val newSyncMap: SyncMap = currentSyncs - dcInfo.key + (dcInfo.key -> SyncerDone(dcInfo.positionKey.get))
         currentSyncs = newSyncMap
+        if(dcInfo.dcInfoExtraType == "fingerprint")
+          zStore.putString("fp-position", dcInfo.positionKey.get)
```
Comment (e-orz) [inline server/cmwell-dc/src/main/scala/cmwell/dc/stream/DataCenterSyncManager.scala]: Add a log print in case persisting in zStore fails.

**Pull request merged**